jsonschema/exceptions.py

```diff
@@ -301,8 +341,31 @@ def by_relevance(weak=WEAK_MATCHES, strong=STRONG_MATCHES):
             a collection of validator names to consider to be "strong"
     """
     def relevance(error):
-        validator = error.validator
-        return -len(error.path), validator not in weak, validator in strong
+        same_type = False
```

Comment:
So! Thanks again for trying to tackle this.

I still want to go for something a lot simpler than the code here. Maybe this will help as a start -- I think the changes to this file and this function probably should be just something like:

```
        same_type = False
        if "type" in error.schema:
            same_type = error._type_checker.is_type(error.instance, error.schema["type"])
        validator = error.validator
        return (
            -len(error.path),
            error.validator not in weak,
            error.validator in strong,
            not same_type,
        )
```

The missing piece there is that the type checker that's used is not in fact carried onto Errors at the minute, but that should be 1-2 lines to add.

If you want to see what that does, you can try instead starting with:

```
        same_type = False
        if "type" in error.schema:
            same_type = _types.draft7_type_checker.is_type(error.instance, error.schema["type"])
        validator = error.validator
        return (
            -len(error.path),
            error.validator not in weak,
            error.validator in strong,
            not same_type,
        )
```

And looking to see what it does to your test cases. From here it looks like it passes your first new one (which was the one that #728 was mainly targeting).

It fails the other two, but I'm not 100% convinced on the other two personally. If we do want to support them though, which yeah we should think about, there are other ways to handle them -- e.g. perhaps errors from non-container types (simple value types, e.g. strings, bools, etc.) should prioritize enum and const validators.

Hopefully the above is a breadcrumb to move this further a bit?

Let me know either way, and probably also worth having a look at CI to make sure it passes with whatever change, it'll be needed before we merge.

Appreciated again for staying on top of this!